pallets/erc20-xcm-bridge/src/lib.rs

```diff
@@ -0,0 +1,184 @@
+			ensure!(
+				!exec_info.value.is_empty() && exec_info.value == bytes,
+				Erc20TransferError::ContractReturnInvalidValue
+			);
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> xcm_executor::traits::TransactAsset for Pallet<T> {
+		// For optimization reasons, the asset we want to deposit has not really been withdrawn,
+		// we have just traced from which account it should have been withdrawn.
+		// So we will retrieve these information and make the transfer from the origin account.
+		fn deposit_asset(what: &MultiAsset, who: &MultiLocation) -> XcmResult {
+			let Erc20Asset {
+				contract_address,
+				amount,
+				maybe_holder,
+			} = Erc20Matcher::<T>::matches_erc20(what)?;
+
+			let from = maybe_holder.ok_or(MatchError::AssetIdConversionFailed)?;
+
+			let beneficiary = T::AccountIdConverter::convert_ref(who)
+				.map_err(|()| MatchError::AccountIdConversionFailed)?;
+
+			frame_support::storage::with_storage_layer(|| {
+				Self::erc20_transfer(contract_address, from, beneficiary, amount)
+					.map_err(DepositError::Erc20TransferError)
+			})
+			.map_err(Into::into)
+		}
+
+		fn internal_transfer_asset(
+			asset: &MultiAsset,
+			from: &MultiLocation,
+			to: &MultiLocation,
+		) -> Result<Assets, XcmError> {
+			let Erc20Asset {
+				contract_address,
+				amount,
+				maybe_holder,
+			} = Erc20Matcher::<T>::matches_erc20(asset)?;
+
+			ensure!(
+				maybe_holder.is_none(),
+				XcmError::from(MatchError::AssetIdConversionFailed)
+			);
+
+			let from = T::AccountIdConverter::convert_ref(from)
+				.map_err(|()| MatchError::AccountIdConversionFailed)?;
+
+			let to = T::AccountIdConverter::convert_ref(to)
+				.map_err(|()| MatchError::AccountIdConversionFailed)?;
+
+			Self::erc20_transfer(contract_address, from, to, amount)?;
+
+			Ok(asset.clone().into())
+		}
+
+		// Since we don't control the erc20 contract that manages the asset we want to withdraw,
+		// we can't really withdraw this asset, we can only transfer it to another account.
+		// It would be possible to transfer the asset to a dedicated account that would reflect
+		// the content of the xcm holding, but this would imply to perform two evm calls instead of
+		// one (1 to withdraw the asset and a second one to deposit it).
+		// In order to perform only one evm call, we just trace the origin of the asset,
+		// and then the transfer will only really be performed in the deposit instruction.
+		fn withdraw_asset(what: &MultiAsset, who: &MultiLocation) -> Result<Assets, XcmError> {
+			let Erc20Asset {
+				contract_address: _,
+				amount: _,
+				maybe_holder,
+			} = Erc20Matcher::<T>::matches_erc20(what)?;
+
+			ensure!(
+				maybe_holder.is_none(),
+				XcmError::from(MatchError::AssetIdConversionFailed)
+			);
+
+			let who = T::AccountIdConverter::convert_ref(who)
+				.map_err(|()| MatchError::AccountIdConversionFailed)?;
+
+			Ok(Erc20Matcher::<T>::insert_holder(what, who)?.into())
+		}
+	}
+}
```

Comment:
To be evaluated whether `XcmContext` in `xcmV3` could contain original_origin. This would allow us to not do this holder insertion in a structure that is not supposed to contain this information like a `MultiAsset`.

Reply:
This actually deviates a lot from the XCM language. An example:

I do a `WithdrawAsset` of an ERC20 token
I do a `DepositAsset` of the same exact token

If I want a Definite `DepositAsset` I actually need to pass a X2 Junctions with the holder on it...Not sure how this will hold over time

Reply:
I reverted this wrong idea and followed back to environmental